File: api/app/oauth/service.py
from __future__ import annotations

import base64
import hashlib
import hmac
import secrets
from collections.abc import Iterable
from datetime import UTC, datetime
from uuid import uuid4

from fastapi import Request

from ..config import Settings
from .models import OAuthAuthorizationCodeRecord, OAuthClientRecord

_DEFAULT_SCOPE = "mcp:read"
_SCOPE_ALIAS_MAP = {
    "read": "mcp:read",
    "write": "mcp:write",
}


def issuer_url_for_request(*, request: Request, settings: Settings) -> str:
    configured = (settings.oauth_issuer_url or "").strip()
    if configured:
        return configured.rstrip("/")
    base_url = str(request.base_url).strip()
    return base_url.rstrip("/")


def normalize_scope(scope: str | None) -> str:
    tokens = [(item or "").strip() for item in (scope or "").split(" ")]
    normalized: list[str] = []
    seen: set[str] = set()
    for token in tokens:
        if not token:
            continue
        mapped = _SCOPE_ALIAS_MAP.get(token, token)
        if mapped in seen:
            continue
        seen.add(mapped)
        normalized.append(mapped)

    if not normalized:
        return _DEFAULT_SCOPE
    return " ".join(normalized)


def token_scope_from_oauth_scope(scope: str | None) -> str:
    normalized = normalize_scope(scope)
    tokens = set(normalized.split())
    if "mcp:write" in tokens:
        return "write"
    return "read"


def oauth_secret_hash(*, identifier: str, secret: str, pepper: str) -> str:
    payload = f"{identifier}:{secret}".encode()
    return hmac.new(pepper.encode(), payload, hashlib.sha256).hexdigest()


def verify_oauth_secret(
    *,
    identifier: str,
    secret: str,
    pepper: str,
    expected_hash: str,
) -> bool:
    candidate = oauth_secret_hash(identifier=identifier, secret=secret, pepper=pepper)
    return hmac.compare_digest(candidate, expected_hash)


def build_oauth_client_id() -> str:
    return f"engram_client_{uuid4().hex}"


def issue_oauth_client_secret() -> str:
    return secrets.token_urlsafe(32)


def generate_authorization_code() -> str:
    return secrets.token_urlsafe(48)


def authorization_code_hash(*, code: str, pepper: str) -> str:
    return oauth_secret_hash(identifier="authorization_code", secret=code, pepper=pepper)


def authorization_code_is_active(
    record: OAuthAuthorizationCodeRecord,
    *,
    now: datetime | None = None,
) -> bool:
    current = now or datetime.now(UTC)
    if record.consumed_at is not None:
        return False
    return record.expires_at > current


def redirect_uri_allowed(*, redirect_uri: str, allowed_uris: Iterable[str]) -> bool:
    candidate = redirect_uri.strip()
    if not candidate:
        return False
    allowed = {item.strip() for item in allowed_uris if item and item.strip()}
    return candidate in allowed


def validate_pkce(
    *,
    code_verifier: str,
    code_challenge: str,
    code_challenge_method: str,
) -> bool:
    method = (code_challenge_method or "S256").strip()
    verifier = (code_verifier or "").strip()
    challenge = (code_challenge or "").strip()
    if not verifier or not challenge:
        return False
    if method == "S256":
        digest = hashlib.sha256(verifier.encode()).digest()
        encoded = base64.urlsafe_b64encode(digest).decode().rstrip("=")
        return hmac.compare_digest(encoded, challenge)
    return False


def client_supports_authorization_code(record: OAuthClientRecord) -> bool:
    return "authorization_code" in set(record.grant_types)

File: internal/oauth/service.go
package oauth

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"net/http"
	"strings"
	"time"

	"engram/internal/config"
	"engram/internal/models"

	"github.com/google/uuid"
)

const (
	defaultScope = "mcp:read"
)

var scopeAliasMap = map[string]string{
	"read":  "mcp:read",
	"write": "mcp:write",
}

// SecretHashInput captures hash inputs for OAuth secrets.
type SecretHashInput struct {
	Identifier string
	Secret     string
	Pepper     string
}

// SecretVerificationInput captures hash verification inputs for OAuth secrets.
type SecretVerificationInput struct {
	Identifier   string
	Secret       string
	Pepper       string
	ExpectedHash string
}

// AuthorizationCodeHashInput captures authorization code hash values.
type AuthorizationCodeHashInput struct {
	Code   string
	Pepper string
}

// RedirectURIAllowedInput captures redirect URI allowlist checks.
type RedirectURIAllowedInput struct {
	RedirectURI string
	AllowedURIs []string
}

// PKCEValidationInput captures PKCE verifier/challenge values.
type PKCEValidationInput struct {
	CodeVerifier        string
	CodeChallenge       string
	CodeChallengeMethod string
}

// IssuerURLForRequest resolves issuer URL using explicit config or request-derived origin.
func IssuerURLForRequest(request *http.Request, settings config.Settings) string {
	configured := strings.TrimSpace(settings.OAuthIssuerURL)
	if configured != "" {
		return strings.TrimRight(configured, "/")
	}
	if request == nil {
		return ""
	}
	scheme := "http"
	if request.TLS != nil {
		scheme = "https"
	}
	return strings.TrimRight(scheme+"://"+request.Host, "/")
}

// NormalizeScope aliases and deduplicates scope values, defaulting to mcp:read.
func NormalizeScope(scope string) string {
	tokens := strings.Split(scope, " ")
	normalized := make([]string, 0, len(tokens))
	seen := make(map[string]struct{}, len(tokens))
	for _, raw := range tokens {
		token := strings.TrimSpace(raw)
		if token == "" {
			continue
		}
		mapped := mapScopeAlias(token)
		if _, exists := seen[mapped]; exists {
			continue
		}
		seen[mapped] = struct{}{}
		normalized = append(normalized, mapped)
	}
	if len(normalized) == 0 {
		return defaultScope
	}
	return strings.Join(normalized, " ")
}

func mapScopeAlias(token string) string {
	if mapped, ok := scopeAliasMap[token]; ok {
		return mapped
	}
	return token
}

// TokenScopeFromOAuthScope maps OAuth scope tokens into MCP token scope.
func TokenScopeFromOAuthScope(scope string) string {
	normalized := NormalizeScope(scope)
	tokens := strings.Fields(normalized)
	for _, token := range tokens {
		if token == "mcp:write" {
			return "write"
		}
	}
	return "read"
}

// OAuthSecretHash returns a peppered HMAC-SHA256 hash for OAuth secrets.
func OAuthSecretHash(input SecretHashInput) string {
	payload := input.Identifier + ":" + input.Secret
	mac := hmac.New(sha256.New, []byte(input.Pepper))
	_, _ = mac.Write([]byte(payload))
	return hex.EncodeToString(mac.Sum(nil))
}

// VerifyOAuthSecret compares candidate and expected secret hashes in constant time.
func VerifyOAuthSecret(input SecretVerificationInput) bool {
	candidate := OAuthSecretHash(SecretHashInput{
		Identifier: input.Identifier,
		Secret:     input.Secret,
		Pepper:     input.Pepper,
	})
	return hmac.Equal([]byte(candidate), []byte(input.ExpectedHash))
}

// BuildOAuthClientID creates a new client_id value.
func BuildOAuthClientID() string {
	compactUUID := strings.ReplaceAll(uuid.NewString(), "-", "")
	return "engram_client_" + compactUUID
}

// IssueOAuthClientSecret returns a URL-safe random client secret.
func IssueOAuthClientSecret() (string, error) {
	return issueURLSafeToken(32)
}

// GenerateAuthorizationCode returns a URL-safe random authorization code.
func GenerateAuthorizationCode() (string, error) {
	return issueURLSafeToken(48)
}

func issueURLSafeToken(byteLength int) (string, error) {
	buffer := make([]byte, byteLength)
	if _, err := rand.Read(buffer); err != nil {
		return "", err
	}
	return base64.RawURLEncoding.EncodeToString(buffer), nil
}

// AuthorizationCodeHash returns a peppered hash for authorization code storage.
func AuthorizationCodeHash(input AuthorizationCodeHashInput) string {
	return OAuthSecretHash(SecretHashInput{
		Identifier: "authorization_code",
		Secret:     input.Code,
		Pepper:     input.Pepper,
	})
}

// AuthorizationCodeIsActive reports whether an authorization code can still be used.
func AuthorizationCodeIsActive(record models.OAuthAuthorizationCodeRecord, now time.Time) bool {
	if record.ConsumedAt != nil {
		return false
	}
	return record.ExpiresAt.After(now)
}

// RedirectURIAllowed reports whether a redirect URI is present in the allowlist.
func RedirectURIAllowed(input RedirectURIAllowedInput) bool {
	candidate := strings.TrimSpace(input.RedirectURI)
	if candidate == "" {
		return false
	}
	for _, rawAllowedURI := range input.AllowedURIs {
		allowedURI := strings.TrimSpace(rawAllowedURI)
		if allowedURI == "" {
			continue
		}
		if candidate == allowedURI {
			return true
		}
	}
	return false
}

// ValidatePKCE validates an S256 PKCE verifier/challenge pair.
func ValidatePKCE(input PKCEValidationInput) bool {
	method := strings.TrimSpace(input.CodeChallengeMethod)
	if method == "" {
		method = "S256"
	}
	verifier := strings.TrimSpace(input.CodeVerifier)
	challenge := strings.TrimSpace(input.CodeChallenge)
	if verifier == "" || challenge == "" {
		return false
	}
	if method != "S256" {
		return false
	}
	digest := sha256.Sum256([]byte(verifier))
	encoded := base64.RawURLEncoding.EncodeToString(digest[:])
	return hmac.Equal([]byte(encoded), []byte(challenge))
}

// ClientSupportsAuthorizationCode reports whether the client allows authorization_code grant.
func ClientSupportsAuthorizationCode(record models.OAuthClientRecord) bool {
	for _, grantType := range record.GrantTypes {
		if grantType == "authorization_code" {
			return true
		}
	}
	return false
}
